File: hash-table/leetcode-242/solution.cc
#include <string>
#include <map>

using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        if(s.length() != t.length()) {
            return false;
        }

        map<int, int> hashMap;
        map<int, int>::iterator itr;
        for(int i = 0; i < s.length(); i++) {
            itr = hashMap.find(s[i]);
            if(hashMap.end() != itr) {
                itr->second++;
            } else {
                hashMap.insert(pair<int, int>(s[i], 1));
            }
        }

        for(int i = 0; i < t.length(); i++) {
            itr = hashMap.find(s[i]);
            if(hashMap.end() != itr) {
                itr->second--;
            } else {
                return false;
            }
        }

        for (itr = hashMap.begin(); itr != hashMap.end(); ++itr) {
            if(0 != itr->second) {
                return false;
            }
        }

        return true;
    }
};

File: hash-table/leetcode-242/solution.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()) {
            return false;
        }

        Map<Character, Integer> hashMap = new HashMap<Character, Integer>();
        for(int i = 0; i < s.length(); i++) {
            if(hashMap.containsKey(s.charAt(i))) {
                hashMap.put(s.charAt(i), hashMap.get(s.charAt(i))+1);
            } else {
                hashMap.put(s.charAt(i), 1);
            }
        }

        for(int i = 0; i < t.length(); i++) {
            if(hashMap.containsKey(t.charAt(i))) {
                hashMap.put(t.charAt(i), hashMap.get(t.charAt(i))-1);
            } else {
                return false;
            }
        }

        for(Map.Entry<Character, Integer> entry : hashMap.entrySet()){
            Integer mapValue = entry.getValue();
            if(0 != mapValue) {
                return false;
            }
        }

        return true;
    }
}